src/canisters/frontend/ic-certified-assets/src/state_machine.rs

```diff
@@ -36,12 +44,192 @@ pub struct AssetEncoding {
+pub struct AssetRedirect {
+    pub from: Option<RedirectUrl>,
+    pub to: RedirectUrl,
+    pub user_agent: Option<Vec<String>>,
+    pub response_code: u16,
+}
+
+impl AssetRedirect {
+    fn get_location_url(&self, req: &HttpRequest) -> Option<RedirectUrl> {
+        fn match_and_replace_segment(
+            replace_from: &Option<String>,
+            replace_with: &Option<String>,
+            replace_in: &str,
+        ) -> Option<String> {
+            if let Some(with) = replace_with {
+                if replace_from.is_none() {
+                    return Some(with.to_owned());
+                }
+                if let Ok(re) = Regex::new(replace_from.as_ref().unwrap()) {
+                    return re
+                        .find(replace_in)
+                        .map(|_| re.replace(replace_in, with).into());
+                }
+            }
+            None
+        }
+
+        let mut location = RedirectUrl::new(None, None);
+
+        if let Some(RedirectUrl {
+            host: from_host,
+            path: from_path,
+        }) = &self.from
+        {
+            let (mut host_same_as_original, mut path_same_as_original) = (true, true);
+            if let Some(req_host) = req.get_header_value("Host") {
+                if let Some(redirect_host) =
+                    match_and_replace_segment(&from_host, &self.to.host, req_host)
+                {
+                    host_same_as_original = &redirect_host == req_host;
+                    location.host = Some(redirect_host);
+                }
+            }
+
+            if let Some(redirect_path) =
+                match_and_replace_segment(&from_path, &self.to.path, &req.url)
+            {
+                path_same_as_original = redirect_path == req.url;
+                location.path = Some(redirect_path);
+            } else {
+                location.path = Some(req.url.clone());
+            }
+
+            if host_same_as_original && path_same_as_original {
+                #[cfg(feature = "local_debugging")]
+                ic_cdk::print(format!(
+                    "redirect loop detected: req:{:?} => from:{:?} -> to:{:?}",
+                    req, self.from, self.to
+                ));
+                return None;
+            }
+        } else {
+            if let Some(host) = &self.to.host {
+                if host.contains("http") {
+                    location.host = Some(host.to_string());
+                } else {
+                    location.host = Some(format!("https://{}", host));
+                }
+            }
+            if let Some(path) = &self.to.path {
+                location.path = Some(path.to_string());
+            } else {
+                location.path = Some(req.url.clone());
+            }
+        }
+
+        Some(location)
+    }
+
+    fn redirect(&self, req: &HttpRequest) -> Option<HttpResponse> {
+        if let Err(e) = self.is_valid() {
+            return Some(HttpResponse::build_400(&e));
+        }
```

Comment:
I think it's too late to check validity at this point. Ideally, we should validate that the configuration is consistent on batch commit.

Reply:
validity is also being checked during batch commit
https://github.com/dfinity/sdk/pull/2414/files#diff-b24a9df9fbc0c1766d0396a9748c7b50b8dc4cebacb9f7230d42d096ed16754fR31-R65

Somewhere around this or next sprint, we will also add support for CRUD methods for properties for each asset in the asset canister. The line you've reviewed is a safety measure if someone were to attempt to change the properties of the asset and set an invalid redirect configuration; however, since there is no official support for mentioned CRUD methods, maybe it's a good idea to move this into "CRUD" PR - is the idea sound to you?